lib/python/treadmill_aws/ipakeytab.py

```diff
@@ -22,151 +25,258 @@
+                    'Hostname in request [%s] does not match hostname in '
+                    'requestor principal [%s]' % (request, requestor))
+
+        def _authorize_user_request(self, requestor, request):
+            requestor_name, requestor_inst, requestor_realm = \
+                _parse_name(requestor)
+            request_name, request_inst, request_realm = _parse_name(request)
 
-        def _get_keytab(self, name, inst, realm):
+            if requestor_name != request_name:
+                raise ValueError(
+                    'Requestor [%s] is not authorized to request keytab for '
+                    '[%s]' % (requestor, request))
+
+            if requestor_realm != self.realm:
+                raise ValueError(
+                    'Requestor [%s] is not in authorized realm' % (requestor))
+
+        def _authorize_admin(self, requestor):
+
+            if admin_group is None:
+                return False
+
+            requestor_name, requestor_inst, requestor_realm = \
+                _parse_name(requestor)
+
+            try:
+                if requestor_name in grp.getgrnam(self.admin_group).gr_mem \
+                        and requestor_inst == '' \
+                        and requestor_realm == self.realm:
+                    return True
+            except KeyError as err:
+                _LOGGER.warning(
+                    'admin group [%s] does not exist', self.admin_group)
+
+            return False
+
+        def _authorize(self, requestor, request):
+
+            # admin can request anything
+            if self._authorize_admin(requestor):
+                return
+
+            # not admin, authorize based on request
+            if '/' in request:
+                # request is for service principal keytab
+                self._authorize_service_request(requestor, request)
+            else:
+                # request is for user keytab
+                self._authorize_user_request(requestor, request)
+
+        def _get_keytab_entries(self, request):
+            name, inst, realm = _parse_name(request)
             tmpktdir = tempfile.mkdtemp(prefix="ipakeytab-")
             tmpkt = "%s/%s" % (tmpktdir, name)
-            service = "%s/%s@%s" % (name, inst, realm)
 
             _LOGGER.info(
-                "writing keytab for %s/%s@%s => %s",
-                name,
-                inst,
-                realm,
-                tmpkt)
+                "writing keytab for [%s] to tmp keytab file %s",
+                request, tmpkt)
 
             os.system(
                 "kadmin.local ktadd -k %s -norandkey %s" %
-                (tmpkt, service))
-            if not os.path.isfile(tmpkt):
-                os.system("klist -5 > /tmp/klist.out")
-                _LOGGER.info("running: ipa service-add %s", service)
-                os.system("ipa service-add %s" % service)
-                os.system("kadmin.local cpw -randkey %s" % service)
+                (tmpkt, request))
```

Comment:
rather than os.system, use 

from treadmill import subproc

subproc.check_call('klist', '-s', ...])

for subproc to work, you need to add certain exes to aliases:

https://github.com/andreikeis/treadmill-aws/blob/master/lib/python/treadmill_aws/bootstrap/aws_aliases.py

aliases are map of symbolic name to full path. notice that you can't have . in symbolic name, so it will be

kadmin_local: /usr/bin/kadmin.local

and subproc.check_call(['kardmin_local', ...])